toolsrc/src/tls12-download.c

```diff
@@ -0,0 +1,269 @@
+ * unavailable to PowerShell.
+ * To that end it avoids using C runtime functions (beyond the vcruntime ones the compiler
+ * injects itself).
+ * (In testing as of 2021-01-07, this version that doesn't link with the CRT is ~8kb, whereas a
+ * hello world program that does link with the CRT is ~300kb)
+ */
+
+void __declspec(noreturn) win32_abort() {
+    /*
+     * Note that TerminateProcess does not return when called from the terminated process, see
+     * https://github.com/MicrosoftDocs/sdk-api/pull/626
+     */
+    TerminateProcess(GetCurrentProcess(), 3);
+    }
+
+size_t wide_length(const wchar_t* str)
+{
+    size_t answer = 0;
+    while (*str)
+    {
+        ++answer;
+        ++str;
+    }
+    return answer;
+}
+
+void write_message(const HANDLE hStdOut, const wchar_t* msg)
+{
+    size_t wcharsToWrite = wide_length(msg);
+    if (wcharsToWrite == 0)
+    {
+        return;
+    }
+
+    if (wcharsToWrite > 65535)
+    {
+        win32_abort();
+    }
+
+    if (WriteConsoleW(hStdOut, msg, wcharsToWrite, 0, 0))
+    {
+        return;
+    }
+
+    // this happens if output has been redirected
+    int narrowChars = WideCharToMultiByte(CP_ACP, 0, msg, (int)wcharsToWrite, 0, 0, 0, 0);
+    if (narrowChars == 0)
+    {
+        win32_abort();
+    }
+
+    char* narrowBuffer = HeapAlloc(GetProcessHeap(), 0, (size_t)narrowChars);
+    if (WideCharToMultiByte(CP_ACP, 0, msg, (int)wcharsToWrite, narrowBuffer, narrowChars, 0, 0) == 0)
+    {
+        win32_abort();
+    }
+
+    while (narrowChars != 0)
+    {
+        DWORD charsWritten;
+        if (!WriteFile(hStdOut, narrowBuffer, (DWORD)narrowChars, &charsWritten, 0))
+        {
+            win32_abort();
+        }
+
+        narrowChars -= (int)charsWritten;
+    }
+
+    if (!HeapFree(GetProcessHeap(), 0, narrowBuffer))
+    {
+        win32_abort();
+    }
+}
+
+void write_number(const HANDLE hStdOut, DWORD number)
+{
+    wchar_t buffer[11]; // 4294967295\0
+    wchar_t* firstDigit = buffer + 11;
+    *--firstDigit = L'\0';
+    if (number == 0)
+    {
+        *--firstDigit = L'0';
+    }
+    else
+    {
+        do
+        {
+            *--firstDigit = L'0' + number % 10;
+            number /= 10;
+        } while (number != 0);
+    }
+
+    write_message(hStdOut, firstDigit);
+}
+
+void __declspec(noreturn) abort_api_failure(const HANDLE hStdOut, const wchar_t* api_name)
+{
+    DWORD lastError = GetLastError();
+    write_message(hStdOut, L"While calling Windows API function ");
+    write_message(hStdOut, api_name);
+    write_message(hStdOut, L"\r\n");
+    wchar_t* message;
+    if (FormatMessageW(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_FROM_HMODULE | FORMAT_MESSAGE_ALLOCATE_BUFFER,
+                       GetModuleHandleW(L"winhttp.dll"),
+                       lastError,
+                       0,
+                       (LPWSTR)&message,
+                       0,
+                       0))

```

Comment:
```suggestion
                       NULL))
```

afaict from the docs (having to read them from markdown is hard... I think they screwed up generation or something...), this is a va_list*; best to use `NULL` explicitly for pointer parameters.

Reply:
I don't want to say `NULL` because I need `<stdlib.h>` to say that and I'm avoiding linking with the CRT here.